core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -0,0 +1,398 @@
+            transactionStartTime = time.milliseconds(),
+            entryTimestamp = time.milliseconds())
+
+          val metadata = txnManager.addTransaction(transactionalId, newMetadata)
+
+          // there might be a concurrent thread that has just updated the mapping
+          // with the transactional id at the same time; in this case we will
+          // treat it as the metadata has existed and update it accordingly
+          metadata synchronized {
+            if (!metadata.equals(newMetadata))
+              metadata.producerEpoch = (metadata.producerEpoch + 1).toShort
+
+            appendMetadataToLog(transactionalId,  metadata, responseCallback)
+          }
+        case Some(metadata) =>
+          initPidWithExistingMetadata(transactionalId, transactionTimeoutMs, responseCallback, metadata)
+      }
+    }
+  }
+
+  private def appendMetadataToLog(transactionalId: String,
+                             metadata: TransactionMetadata,
+                             initPidCallback: InitPidCallback): Unit ={
+    def callback(errors: Errors): Unit = {
+      if (errors == Errors.NONE) initPidCallback(initTransactionMetadata(metadata))
+      else initPidCallback(initTransactionError(errors))
+    }
+
+    txnManager.appendTransactionToLog(transactionalId, metadata, callback)
+
+  }
+
+
+  private def initPidWithExistingMetadata(transactionalId: String,
+                                          transactionTimeoutMs: Int,
+                                          responseCallback: InitPidCallback,
+                                          metadata: TransactionMetadata) = {
+
+    metadata synchronized {
+      if (metadata.state == Ongoing) {
+        // abort the ongoing transaction
+        handleEndTransaction(transactionalId,
+          metadata.pid,
+          metadata.producerEpoch,
+          TransactionResult.ABORT,
+          (errors: Errors) => {
+            if (errors != Errors.NONE) {
+              responseCallback(initTransactionError(errors))
+            } else {
+              // init pid again
+              handleInitPid(transactionalId, transactionTimeoutMs, responseCallback)
+            }
+          })
+      } else if (metadata.state == PrepareAbort || metadata.state == PrepareCommit) {
+        // wait for the commit to complete and then init pid again
+        txnMarkerPurgatory.tryCompleteElseWatch(new DelayedTxnMarker(metadata, (errors: Errors) => {
+          if (errors != Errors.NONE)
+            responseCallback(initTransactionError(errors))
+          else
+            handleInitPid(transactionalId, transactionTimeoutMs, responseCallback)
+        }), Seq(metadata.pid))
+      } else {
+        metadata.producerEpoch = (metadata.producerEpoch + 1).toShort
+        metadata.txnTimeoutMs = transactionTimeoutMs
+        metadata.topicPartitions.clear()
+        metadata.entryTimestamp = time.milliseconds()
+        metadata.transactionStartTime = time.milliseconds()
```

Comment:
We can set it to -1, but it may not matter. transactionStartTime is only useful when aborting a transaction that's started. If the transaction is in empty state, transactionStartTime is probably not going to be used. It's more important to set transactionStartTime on adding the first partition.